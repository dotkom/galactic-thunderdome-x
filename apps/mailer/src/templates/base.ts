import { type Template } from "../template-service"

export type BaseTemplateProps = {
  body: string
}

export const template: Template<BaseTemplateProps> = (props) => `
<table width="100%" cellpadding="0" cellspacing="0" align="center">
  <tr>
    <td width="100%" align="center">
      <!-- Main content -->
      <table width="600" cellpadding="0" cellspacing="0" style="background-color: #FFF; padding-bottom: 20px;" align="center">
        <tr>
          <td>
            ${props.body}
          </td>
        </tr>
      </table>
        <!-- Footer -->
      <table width="600" cellpadding="0" cellspacing="0" style="background-color: #000; color: #FFF;" align="center">
        <tr style="height: 30px;">
          <td align="center">
            <img style="margin-top: 5px;" width="220" src="https://onlineweb4-prod.s3.eu-north-1.amazonaws.com/media/wiki/attachments/70/52fda8303327a69819bc245931869b51/Online_hvit.png" />
          </td>
        </tr>
        <tr>
          <td>
            <p style="font-size: 13px; font-family: sans-serif; text-align: center;">Linjeforeningen Online (OrgNr 992 548 04), Sem SÃ¦lands vei 9, 7491 Trondheim</p>
          </td>
        </tr>
      </table>
    </td>
  </tr>
</table>
`
