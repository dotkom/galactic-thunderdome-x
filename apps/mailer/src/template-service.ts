import { template as baseTemplate } from "./templates/base"
import { marked } from "marked"

// Mapping of templates to their respective functions
export const templates = {
  base: baseTemplate,
} as const

export interface TemplateService {
  /** Render a named template with that template's arguments into HTML */
  render<K extends keyof TemplateMap>(name: K, props: TemplateMap[K]): string

  /** Transform Markdown into HTML */
  transform(markdown: string): string
}

export const initTemplateService = (): TemplateService => {
  return {
    render: <K extends keyof TemplateMap>(name: K, props: TemplateMap[K]) => {
      return templates[name](props)
    },
    transform: (markdown: string): string => {
      return marked.parse(markdown)
    },
  }
}

export type Template<T> = (props: T) => string
export type TemplateMap = {
  [K in keyof typeof templates]: typeof templates[K] extends Template<infer U> ? U : never
}
