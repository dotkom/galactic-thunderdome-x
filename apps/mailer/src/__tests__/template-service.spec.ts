import { initTemplateService } from "../template-service"
import { BaseTemplateProps } from "../templates/base"

describe("TemplateService", () => {
  const templateService = initTemplateService()

  it("should interpolate the properties into the template", () => {
    const props: BaseTemplateProps = { body: "Hello world" }
    const template = templateService.render("base", props)
    expect(template).toMatchSnapshot()
  })
})
