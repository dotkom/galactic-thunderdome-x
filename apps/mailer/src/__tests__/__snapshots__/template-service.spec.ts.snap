// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateService should be able to transform markdown into html 1`] = `
"<h1 id=\\"marked---markdown-parser\\">Marked - Markdown Parser</h1>
<p><a href=\\"https://github.com/markedjs/marked/\\">Marked</a> lets you convert <a href=\\"http://daringfireball.net/projects/markdown/\\">Markdown</a> into HTML.  Markdown is a simple text format whose goal is to be very easy to read and write, even when not converted to HTML.  This demo page will let you type anything you like and see how it gets converted.  Live.  No more waiting around.</p>
<h2 id=\\"how-to-use-the-demo\\">How To Use The Demo</h2>
<ol>
<li>Type in stuff on the left.</li>
<li>See the live updates on the right.</li>
</ol>
<p>That&#39;s it.  Pretty simple.  There&#39;s also a drop-down option in the upper right to switch between various views:</p>
<ul>
<li><strong>Preview:</strong>  A live display of the generated HTML as it would render in a browser.</li>
<li><strong>HTML Source:</strong>  The generated HTML before your browser makes it pretty.</li>
<li><strong>Lexer Data:</strong>  What <a href=\\"https://github.com/markedjs/marked/\\">marked</a> uses internally, in case you like gory stuff like this.</li>
<li><strong>Quick Reference:</strong>  A brief run-down of how to format things using markdown.</li>
</ul>
<h2 id=\\"why-markdown\\">Why Markdown?</h2>
<p>It&#39;s easy.  It&#39;s not overly bloated, unlike HTML.  Also, as the creator of <a href=\\"http://daringfireball.net/projects/markdown/\\">markdown</a> says,</p>
<blockquote>
<p>The overriding design goal for Markdown&#39;s
formatting syntax is to make it as readable
as possible. The idea is that a
Markdown-formatted document should be
publishable as-is, as plain text, without
looking like it&#39;s been marked up with tags
or formatting instructions.</p>
</blockquote>
<p>Ready to start writing?  Either start changing stuff on the left or
<a href=\\"/demo/?text=\\">clear everything</a> with a simple click.</p>
"
`;

exports[`TemplateService should interpolate the properties into the template 1`] = `
"
<table width=\\"100%\\" cellpadding=\\"0\\" cellspacing=\\"0\\" align=\\"center\\">
  <tr>
    <td width=\\"100%\\" align=\\"center\\">
      <!-- Main content -->
      <table width=\\"600\\" cellpadding=\\"0\\" cellspacing=\\"0\\" style=\\"background-color: #FFF; padding-bottom: 20px;\\" align=\\"center\\">
        <tr>
          <td>
            Hello world
          </td>
        </tr>
      </table>
        <!-- Footer -->
      <table width=\\"600\\" cellpadding=\\"0\\" cellspacing=\\"0\\" style=\\"background-color: #000; color: #FFF;\\" align=\\"center\\">
        <tr style=\\"height: 30px;\\">
          <td align=\\"center\\">
            <img style=\\"margin-top: 5px;\\" width=\\"220\\" src=\\"https://onlineweb4-prod.s3.eu-north-1.amazonaws.com/media/wiki/attachments/70/52fda8303327a69819bc245931869b51/Online_hvit.png\\" />
          </td>
        </tr>
        <tr>
          <td>
            <p style=\\"font-size: 13px; font-family: sans-serif; text-align: center;\\">Linjeforeningen Online (OrgNr 992 548 04), Sem SÃ¦lands vei 9, 7491 Trondheim</p>
          </td>
        </tr>
      </table>
    </td>
  </tr>
</table>
"
`;
